File: Button.py
import RPi.GPIO as GPIO
import time

class Button:
    button_fell = False
    def __init__(self, pin):
        self.pin = pin
        self.pressed = False
        # setup pin...
        GPIO.setup(pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        GPIO.add_event_detect(pin, GPIO.FALLING, callback=self.button_callback, bouncetime=300)

    def tick(self):
        # button_state = GPIO.input(self.pin)
        # if not button_state:
        #     self.button_fell = True
        #     print("Button pressed")
        #     time.sleep(0.4) # This should be debounced...
        pass # do nothing for now...

    def fell(self):
        if self.button_fell:
            self.button_fell = False
            print("returning true")
            return True
        else:
            return False

    def button_callback(self, channel): # make sure this function works!!
        self.button_fell = True
        print("Button pressed")

File: Button.cpp
/**
 * @file Button.cpp
 *
 * @brief Button class
 *
 * This file implements the Button class, including initialization, button
 * presses, and debouncing
 *
 * @author Bryan Cisneros
 */

#include <bcm2835.h>
#include "Button.h"
#include "Globals.h"
#include <stdio.h>

#define TIMEOUT_COUNT (250)

bool Button::bcm2835_initialized = false;

Button::Button(int pin)
{
    // Make sure the bcm2835 is initialized!
    if(!bcm2835_initialized)
    {
        if (!bcm2835_init())
        {
            printf("bcm2835 init failed!/n");
        }
        bcm2835_initialized = true;
    }

    buttonPin = pin;

    // Set pin as an input
    bcm2835_gpio_fsel(buttonPin, BCM2835_GPIO_FSEL_INPT);

    // Enable pullup on pin
    bcm2835_gpio_set_pud(buttonPin, BCM2835_GPIO_PUD_UP);

    pressed = false;
}


Button::~Button()
{
}

void Button::tick()
{
    // Check for a press, and check that the previous press has already timed out
    if (!bcm2835_gpio_lev(buttonPin) && timeout == 0)
    {
        // Press detected! set the flag, and also set the timeout for debouncing
        pressed = true;
        timeout = TIMEOUT_COUNT;
    }

    // If the timeout is running, decrement it
    if (timeout > 0)
    {
        timeout--;
    }
}

bool Button::fell()
{
    // If the button has been pressed, reset the flag and return true. Otherwise
    // return false;
    if (pressed)
    {
        pressed = false;
        return true;
    }
    else
    {
        return false;
    }

}
